types/types_test.go

```diff
@@ -55,12 +115,148 @@ func TestAlertMerge(t *testing.T) {
 				Timeout:   true,
 			},
 		},
+		{
+			// Both alerts have the Timeout flag unset and are not resolved.
+			// StartsAt is defined by Alert A.
+			// EndsAt is defined by Alert A.
+			A: &Alert{
+				Alert: model.Alert{
+					StartsAt: now.Add(-time.Minute),
+					EndsAt:   now.Add(3 * time.Minute),
+				},
+				UpdatedAt: now,
+			},
+			B: &Alert{
+				Alert: model.Alert{
+					StartsAt: now,
+					EndsAt:   now.Add(2 * time.Minute),
+				},
+				UpdatedAt: now.Add(time.Minute),
+			},
+			Res: &Alert{
+				Alert: model.Alert{
+					StartsAt: now.Add(-time.Minute),
+					EndsAt:   now.Add(3 * time.Minute),
+				},
+				UpdatedAt: now.Add(time.Minute),
+			},
+		},
+		{
+			// Both alerts have the Timeout flag unset and are not resolved.
+			// StartsAt is defined by Alert A.
+			// EndsAt is defined by Alert B.
+			A: &Alert{
+				Alert: model.Alert{
+					StartsAt: now.Add(-time.Minute),
+					EndsAt:   now.Add(3 * time.Minute),
+				},
+				UpdatedAt: now,
+			},
+			B: &Alert{
+				Alert: model.Alert{
+					StartsAt: now.Add(-time.Minute),
+					EndsAt:   now.Add(4 * time.Minute),
+				},
+				UpdatedAt: now.Add(time.Minute),
+			},
+			Res: &Alert{
+				Alert: model.Alert{
+					StartsAt: now.Add(-time.Minute),
+					EndsAt:   now.Add(4 * time.Minute),
+				},
+				UpdatedAt: now.Add(time.Minute),
+			},
+		},
+		{
+			// Both alerts have the Timeout flag unset, A is resolved while B isn't.
+			// StartsAt is defined by Alert A.
+			// EndsAt is defined by Alert B.
+			A: &Alert{
+				Alert: model.Alert{
+					StartsAt: now.Add(-3 * time.Minute),
+					EndsAt:   now.Add(-time.Minute),
+				},
+				UpdatedAt: now,
+			},
+			B: &Alert{
+				Alert: model.Alert{
+					StartsAt: now.Add(-2 * time.Minute),
+					EndsAt:   now.Add(time.Minute),
+				},
+				UpdatedAt: now.Add(time.Minute),
+			},
+			Res: &Alert{
+				Alert: model.Alert{
+					StartsAt: now.Add(-3 * time.Minute),
+					EndsAt:   now.Add(time.Minute),
+				},
+				UpdatedAt: now.Add(time.Minute),
+			},
+		},
+		{
+			// Both alerts have the Timeout flag unset, B is resolved while A isn't.
+			// StartsAt is defined by Alert A.
+			// EndsAt is defined by Alert B.
+			A: &Alert{
+				Alert: model.Alert{
+					StartsAt: now.Add(-2 * time.Minute),
+					EndsAt:   now.Add(3 * time.Minute),
+				},
+				UpdatedAt: now,
+			},
+			B: &Alert{
+				Alert: model.Alert{
+					StartsAt: now.Add(-2 * time.Minute),
+					EndsAt:   now,
+				},
+				UpdatedAt: now.Add(time.Minute),
+			},
+			Res: &Alert{
+				Alert: model.Alert{
+					StartsAt: now.Add(-2 * time.Minute),
+					EndsAt:   now,
```

Comment:
I am confused by this. If I have two alerts, where one is firing and one is not, I expect the merged alert to still be firing, right? Especially as none of them have the Timeout flag set.

Reply:
Because B is received after A, it's EndsAt value "wins" because A isn't resolved (is what I'm reading from the code and comments).

Reply:
> Because B is received after A

Exactly, an alert that is effectively resolved wins over an older alert that was firing previously.

Reply:
Right got it. Thanks for clarifying @stuartnelson3 & @simonpasquier!